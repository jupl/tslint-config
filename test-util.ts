import {readFile} from 'mz/fs'
import {
  Configuration,
  FormatterConstructor,
  IRuleFailureJson,
  Linter,
  RuleFailure,
} from 'tslint'

export interface ILintResult {
  readonly failures: ReadonlyArray<IRuleFailureJson>
  readonly errorCount: number
  readonly warningCount: number
  readonly fixes: ReadonlyArray<IRuleFailureJson>
  readonly format: string | FormatterConstructor
  readonly output: string
}

const basePath = process.cwd()
const configuration = Configuration.loadConfigurationFromPath('./tslint.json')
const linter = new Linter({fix: true}, Linter.createProgram('./tsconfig.json'))

export async function lintFixture(fixtureName: string): Promise<ILintResult> {
  const fileName = `./fixtures/${fixtureName}`
  const file = await readFile(fileName, 'utf8')
  linter.lint(fileName, file, configuration)
  const {failures, output, fixes = [], ...result} = linter.getResult()
  return {
    ...result,
    failures: failures.map(failureToJson),
    fixes: fixes.map(failureToJson),
    output: output.split(basePath).join('.'),
  }
}

function failureToJson(failure: RuleFailure): IRuleFailureJson {
  const {name, ...json} = failure.toJson()
  return {...json, name: name.split(basePath).join('.')}
}
