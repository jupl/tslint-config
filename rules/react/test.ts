import {lintFixture} from '../../test-util'

test('React rules should be valid', async() => {
  const results = await lintFixture('react.tsx')
  expect(results).toMatchSnapshot()
})
