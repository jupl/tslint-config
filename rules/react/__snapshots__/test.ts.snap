// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React rules should be valid 1`] = `
Object {
  "errorCount": 11,
  "failures": Array [
    Object {
      "endPosition": Object {
        "character": 17,
        "line": 4,
        "position": 85,
      },
      "failure": "JSX attributes must be indented further than the opening tag statement",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-alignment",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 2,
        "line": 4,
        "position": 70,
      },
    },
    Object {
      "endPosition": Object {
        "character": 19,
        "line": 5,
        "position": 105,
      },
      "failure": "JSX attributes must be on their own line and vertically aligned",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-alignment",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 4,
        "line": 5,
        "position": 90,
      },
    },
    Object {
      "endPosition": Object {
        "character": 13,
        "line": 28,
        "position": 527,
      },
      "failure": "Closing tag must be on its own line and aligned with opening tag",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-alignment",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 7,
        "line": 28,
        "position": 521,
      },
    },
    Object {
      "endPosition": Object {
        "character": 19,
        "line": 5,
        "position": 105,
      },
      "failure": "Value must be omitted for boolean attributes",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-boolean-value",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 4,
        "line": 5,
        "position": 90,
      },
    },
    Object {
      "endPosition": Object {
        "character": 36,
        "line": 23,
        "position": 422,
      },
      "failure": "Binds are forbidden in JSX attributes due to their rendering performance impact",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-no-bind",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 17,
        "line": 23,
        "position": 403,
      },
    },
    Object {
      "endPosition": Object {
        "character": 32,
        "line": 24,
        "position": 456,
      },
      "failure": "Lambdas are forbidden in JSX attributes due to their rendering performance impact",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-no-lambda",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 16,
        "line": 24,
        "position": 440,
      },
    },
    Object {
      "endPosition": Object {
        "character": 7,
        "line": 28,
        "position": 521,
      },
      "failure": "Multiline JS expressions inside JSX are forbidden",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-no-multiline-js",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 7,
        "line": 26,
        "position": 498,
      },
    },
    Object {
      "endPosition": Object {
        "character": 3,
        "line": 36,
        "position": 589,
      },
      "failure": "Multiline JS expressions inside JSX are forbidden",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-no-multiline-js",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 7,
        "line": 34,
        "position": 574,
      },
    },
    Object {
      "endPosition": Object {
        "character": 25,
        "line": 42,
        "position": 734,
      },
      "failure": "Pass a callback to ref prop instead of a string literal",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-no-string-ref",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 20,
        "line": 42,
        "position": 729,
      },
    },
    Object {
      "endPosition": Object {
        "character": 6,
        "line": 47,
        "position": 788,
      },
      "failure": "JSX elements with no children must be self-closing",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-self-close",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 24,
        "line": 46,
        "position": 776,
      },
    },
    Object {
      "endPosition": Object {
        "character": 6,
        "line": 47,
        "position": 788,
      },
      "failure": "Multiline JSX elements must be wrapped in parentheses",
      "fix": undefined,
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-wrap-multiline",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 24,
        "line": 46,
        "position": 776,
      },
    },
  ],
  "fixes": Array [
    Object {
      "endPosition": Object {
        "character": 5,
        "line": 7,
        "position": 115,
      },
      "failure": "There should be no space after {",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 115,
        "innerText": "",
      },
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-curly-spacing",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 4,
        "line": 7,
        "position": 114,
      },
    },
    Object {
      "endPosition": Object {
        "character": 14,
        "line": 7,
        "position": 124,
      },
      "failure": "There should be no space before }",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 122,
        "innerText": "",
      },
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-curly-spacing",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 13,
        "line": 7,
        "position": 123,
      },
    },
    Object {
      "endPosition": Object {
        "character": 39,
        "line": 11,
        "position": 177,
      },
      "failure": "There should be no space before '='",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 176,
        "innerText": "",
      },
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-equals-spacing",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 38,
        "line": 11,
        "position": 176,
      },
    },
    Object {
      "endPosition": Object {
        "character": 41,
        "line": 11,
        "position": 179,
      },
      "failure": "There should be no space after '='",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 178,
        "innerText": "",
      },
      "name": "./fixtures/react.tsx",
      "ruleName": "jsx-equals-spacing",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 40,
        "line": 11,
        "position": 178,
      },
    },
  ],
  "format": "prose",
  "output": "Fixed 4 error(s) in ./fixtures/react.tsx

ERROR: ./fixtures/react.tsx[5, 3]: JSX attributes must be indented further than the opening tag statement
ERROR: ./fixtures/react.tsx[6, 5]: JSX attributes must be on their own line and vertically aligned
ERROR: ./fixtures/react.tsx[6, 5]: Value must be omitted for boolean attributes
ERROR: ./fixtures/react.tsx[24, 18]: Binds are forbidden in JSX attributes due to their rendering performance impact
ERROR: ./fixtures/react.tsx[25, 17]: Lambdas are forbidden in JSX attributes due to their rendering performance impact
ERROR: ./fixtures/react.tsx[27, 8]: Multiline JS expressions inside JSX are forbidden
ERROR: ./fixtures/react.tsx[29, 8]: Closing tag must be on its own line and aligned with opening tag
ERROR: ./fixtures/react.tsx[35, 8]: Multiline JS expressions inside JSX are forbidden
ERROR: ./fixtures/react.tsx[43, 21]: Pass a callback to ref prop instead of a string literal
ERROR: ./fixtures/react.tsx[47, 25]: JSX elements with no children must be self-closing
ERROR: ./fixtures/react.tsx[47, 25]: Multiline JSX elements must be wrapped in parentheses
",
  "warningCount": 0,
}
`;
